README.md
# dsfa



## Getting started

To make it easy for you to get started with GitLab, here's a list of recommended next steps.

Already a pro? Just edit this README.md and make it your own. Want to make it easy? [Use the template at the bottom](#editing-this-readme)!

## Add your files

- [ ] [Create](https://docs.gitlab.com/ee/user/project/repository/web_editor.html#create-a-file) or [upload](https://docs.gitlab.com/ee/user/project/repository/web_editor.html#upload-a-file) files
- [ ] [Add files using the command line](https://docs.gitlab.com/ee/gitlab-basics/add-file.html#add-a-file-using-the-command-line) or push an existing Git repository with the following command:

```
cd existing_repo
git remote add origin https://gitlab.tu-clausthal.de/rs73/dsfa.git
git branch -M main
git push -uf origin main
```

## Integrate with your tools

- [ ] [Set up project integrations](https://gitlab.tu-clausthal.de/rs73/dsfa/-/settings/integrations)

## Collaborate with your team

- [ ] [Invite team members and collaborators](https://docs.gitlab.com/ee/user/project/members/)
- [ ] [Create a new merge request](https://docs.gitlab.com/ee/user/project/merge_requests/creating_merge_requests.html)
- [ ] [Automatically close issues from merge requests](https://docs.gitlab.com/ee/user/project/issues/managing_issues.html#closing-issues-automatically)
- [ ] [Enable merge request approvals](https://docs.gitlab.com/ee/user/project/merge_requests/approvals/)
- [ ] [Automatically merge when pipeline succeeds](https://docs.gitlab.com/ee/user/project/merge_requests/merge_when_pipeline_succeeds.html)

## Test and Deploy

Use the built-in continuous integration in GitLab.

- [ ] [Get started with GitLab CI/CD](https://docs.gitlab.com/ee/ci/quick_start/index.html)
- [ ] [Analyze your code for known vulnerabilities with Static Application Security Testing(SAST)](https://docs.gitlab.com/ee/user/application_security/sast/)
- [ ] [Deploy to Kubernetes, Amazon EC2, or Amazon ECS using Auto Deploy](https://docs.gitlab.com/ee/topics/autodevops/requirements.html)
- [ ] [Use pull-based deployments for improved Kubernetes management](https://docs.gitlab.com/ee/user/clusters/agent/)
- [ ] [Set up protected environments](https://docs.gitlab.com/ee/ci/environments/protected_environments.html)

***

# Editing this README

When you're ready to make this README your own, just edit this file and use the handy template below (or feel free to structure it however you want - this is just a starting point!). Thank you to [makeareadme.com](https://www.makeareadme.com/) for this template.

## Suggestions for a good README
Every project is different, so consider which of these sections apply to yours. The sections used in the template are suggestions for most open source projects. Also keep in mind that while a README can be too long and detailed, too long is better than too short. If you think your README is too long, consider utilizing another form of documentation rather than cutting out information.

## Name
Choose a self-explaining name for your project.

## Description
Let people know what your project can do specifically. Provide context and add a link to any reference visitors might be unfamiliar with. A list of Features or a Background subsection can also be added here. If there are alternatives to your project, this is a good place to list differentiating factors.

## Badges
On some READMEs, you may see small images that convey metadata, such as whether or not all the tests are passing for the project. You can use Shields to add some to your README. Many services also have instructions for adding a badge.

## Visuals
Depending on what you are making, it can be a good idea to include screenshots or even a video (you'll frequently see GIFs rather than actual videos). Tools like ttygif can help, but check out Asciinema for a more sophisticated method.

## Installation
Within a particular ecosystem, there may be a common way of installing things, such as using Yarn, NuGet, or Homebrew. However, consider the possibility that whoever is reading your README is a novice and would like more guidance. Listing specific steps helps remove ambiguity and gets people to using your project as quickly as possible. If it only runs in a specific context like a particular programming language version or operating system or has dependencies that have to be installed manually, also add a Requirements subsection.

## Usage
Use examples liberally, and show the expected output if you can. It's helpful to have inline the smallest example of usage that you can demonstrate, while providing links to more sophisticated examples if they are too long to reasonably include in the README.

## Support
Tell people where they can go to for help. It can be any combination of an issue tracker, a chat room, an email address, etc.

## Roadmap
If you have ideas for releases in the future, it is a good idea to list them in the README.

## Contributing
State if you are open to contributions and what your requirements are for accepting them.

For people who want to make changes to your project, it's helpful to have some documentation on how to get started. Perhaps there is a script that they should run or some environment variables that they need to set. Make these steps explicit. These instructions could also be useful to your future self.

You can also document commands to lint the code or run tests. These steps help to ensure high code quality and reduce the likelihood that the changes inadvertently break something. Having instructions for running tests is especially helpful if it requires external setup, such as starting a Selenium server for testing in a browser.

## Authors and acknowledgment
Show your appreciation to those who have contributed to the project.

## License
For open source projects, say how it is licensed.

## Project status
If you have run out of energy or time for your project, put a note at the top of the README saying that development has slowed down or stopped completely. Someone may choose to fork your project or volunteer to step in as a maintainer or owner, allowing your project to keep going. You can also make an explicit request for maintainers.


NEWS.md
*   denotes small changes
**  denotes quite substantial/important changes
*** denotes really big changes 
inp "x" -> inp "y" denotes that input argument x has changed to input argument y

## version 2.0.2
* cop_distr -> distr to be more consistent for copulas

* Changed output of elasticity and efficiency function such that confidence intervals are now included.

* comper predict function bug fixed, defined npar.

* Added BurkinaFarms data set, which is used as an example in comper_mv.

* Bounded the inputs for W matrix to be in [0,1]^2 via clamp for more stability

* Improved starting values for the copula parameters via mgcv_cop

* Added "normalized" residuals, which are probability integral transformed residuals which are then transformed via qnorm()

* Introduced a wrapper function for mgcv::gam called "dsfa" in order to be in line with the naming convention of R packages

* Changed link function in comper and comper_mv from log to logshift. This is done for more numerical stability.

## version 2.0.1
* Added rotations 90, 180 and 270 for archimedian copulas and check parameter bounds of delta for dcop, pcop and rcop

* reworked elasticity() output to make it more consistent with efficiency()

* Added cost function example for dsfa and comper

## version 2.0.0
*** Major overhaul of the package.

* dnormhnorm, pnormhnorm, dnormexp, pnormexp, dcop are now wrappers for functions written via RcppArmadillo.
As these are called in the optimization the computation speed could be significantly increased. 

* Reducing dependencies to other packages

* Streamlining naming, comperr -> comper to be in line with erf.

* Adding examples from 'RiceFarms' data set

* Added data set 'manuf'

* Reparametrize is now mom2par and par2mom

## version 1.0.1
* Added rsp function (inverse link function)

* Added pnorm and qnorm functions which handle like the base R function but allow for the derivatives 0 and 2.

* OwenT function has now the option for deriv = 2.

* Added function "check_arguments" to check input arguments for pdf, cdf, quantile function and random number generation of normhnorm and normexp.

** Rewrote "chainrule" function such that the output is a vector with attributes "gradient", "hessian", "l3" and "l4" to be more consistent with the other functions. Further input
   g can be a list now.

* normexp and normhnorm can now take values of the parameters (mu, sigma_v, sigma_u, lambda) which differ in length from x,p,q,n. They are recycled to the appropriate length.

* Added functions "outlier_correct_column" and "outlier_correct" in "utility_functions" which are used to make the calculation of the gradient and hessian of copula more stable.

* Correct loglike of "joe" copula

* Simplified inputs for reparametrize.

* Added postproc for normexp_mgcv, normhnorm_mgcv and comperr_mv_mgcv to add attribute "s" to predicted values, such that qq.gam can now access information on s

* Bivariate normal copula compuation is now faster

* normexp_mgcv, normhnorm_mgcv and comperr_mv_mgcv have now simplified starting values


## version 1.0.0
*** inp"xg" -> inp "tri" in all functions, to be consistent with 'mgcv'

*** dcop, pcop and rcop have been rewritten and allow for independent, clayton, gumbel, frank and joe copula
  inp "U" -> inp "W"
  inp "Tau" -> inp "delta", delta is the copula parameter not kendalls tau anymore
  inp "family" -> inp "family_cop"
  Removed option for disjoint and numerical derivatives.
  Adjusted documentation

* removed cop_utility.R and moved functions into cop.R

*** OwenT function added for less dependencies from "sn" package

*** zeta function added for less dependencies from "sn" package

* erf.R added which calculates error function, complementary error function, inverse error function and inverse complementary error function.
  Further, d1qnormdx1 and d2qnormdx2 were also moved the same way. Formerly these were in utility_functions.R

* Rewrote the help page for 'dsfa'. Added an example.

## version 0.0.3
* initial release version

R/dcop.R
#' Copula function
#'
#' Probablitiy density function, distribution and random number generation for copulas.
#' 
#'
#' @details 
#' A two-dimensional copula \eqn{C(w_1, w_2, \delta)} is a multivariate cumulative distribution function for which the marginal probability distribution of \eqn{w_1} and \eqn{w_2} are uniform on the interval \eqn{[0,1]}.
#' The parameter \eqn{\delta} specifies the copula.
#'
#' The functions \code{pcop()} and \code{rcop()} are wrapper functions for the \code{\link[copula:pCopula]{pCopula}} and \code{\link[copula:rCopula]{rCopula}}.
#'
#' @return \code{dcop} gives the density, \code{pcop} gives the distribution function for a specified copula and \code{rcop} generates random numbers, with given \code{delta}.
#' \code{dcop()} returns a \code{derivs} object. For more details see \code{\link{trind}} and \code{\link{trind_generator}}.
#'
#' @param W numeric matrix of pseudo observations. Must have two columns.
#' @param delta numeric vector of copula parameter \eqn{\delta}.
#' @param distr string, defines the copula family:\cr
#' `independent` = Independence copula \cr
#' `normal` = Gaussian copula \cr
#' `clayton` = Clayton copula \cr
#' `gumbel` = Gumbel copula \cr
#' `frank` = Frank copula \cr
#' `joe` = Joe copula \cr
#' `amh` = Ali-Mikhail-Haq copula \cr
#' @param rot, integer determining the rotation for Archimedian copulas. Can be \code{90}, \code{180} or \code{270}.
#' @inheritParams dcomper
#' 
#' @examples
#' u=0.3; v=0.7; p=0.5
#' pdf <- dcop(W=cbind(u,v), delta=p, distr="normal")
#' cdf <- pcop(W=cbind(u,v), delta=p, distr="normal")
#' r <- rcop(n=100, delta=p, distr="normal")
#'
#' @references
#' \itemize{
#' \item \insertRef{schepsmeier2014derivatives}{dsfa}
#' \item \insertRef{hofert2018elements}{dsfa}
#' }
#' 
#' @family copula
#' 
#' @export
dcop<-function(W, delta, distr="normal", rot=0, deriv_order=0, tri=NULL, log.p=FALSE){
  #Density of copula
  if (any(W <= 0)|any(W >= 1)){
    stop(paste("W must be in [-1, 1]", "\n", ""))
  }
  
  distr<-match.arg(distr,c("independent","normal","clayton","gumbel","frank","joe","amh"))
  
  minmax<-delta_bounds(distr)
  if(any(delta<minmax[1])|any(delta>minmax[2])){
    stop(paste("delta must be in [",minmax[1],",",minmax[2],"]", "\n", ""))
  }
  
  X<-tryCatch(cbind(W, delta), warning=function(w) {
    stop("Input vectors have incompatible lengths")
    })
  
  if(!rot%in%c(0, 90,180,270)){
    stop(paste("rotation must be in {0, 90, 180, 270}", "\n", ""))
  }
  
  if(is.null(tri)){
    tri=trind_generator(3)
  }
  
  out<-dcop_cpp (X[,1], X[,2], X[,3], distr, rot=rot, deriv_order, tri, log.p)
  
  #return out
  return(out)
}

#' @describeIn dcop distribution function for copula.
#' @export
pcop<-function(W, delta=0, distr="normal", rot=0, log.p = FALSE){
  #Distribution function of bivariate copula
  if (any(W <= 0)|any(W >= 1)){
    stop(paste("W must be in [-1, 1]", "\n", ""))
  }
  
  distr<-match.arg(distr,c("independent","normal","clayton","gumbel","frank","joe","amh"))
  
  minmax<-delta_bounds(distr)
  if(any(delta<minmax[1])|any(delta>minmax[2])){
    stop(paste("delta must be in [",minmax[1],",",minmax[2],"]", "\n", ""))
  }
  
  X<-tryCatch(cbind(W, delta), warning=function(w) {
    stop("Input vectors have incompatible lengths")
    })

  if(!rot%in%c(0,90,180,270)){
    stop(paste("rotation must be in {0, 90, 180, 270}", "\n", ""))
  }
  
  out<-sapply(1:nrow(X), function(n) pcop_copula(W=X[n,-3], delta=X[n,3], distr=distr, log.p=log.p))

  #Return out
  return(out)
}

#' @describeIn dcop random number generation for copula.
#' @inheritParams rcomper
#' @export
rcop<-function(n, delta=0, distr="normal", rot=0){
  #Random number generation function of copula
  distr<-match.arg(distr,c("independent","normal","clayton","gumbel","frank","joe","amh"))
  
  minmax<-delta_bounds(distr)
  if(any(delta<minmax[1])|any(delta>minmax[2])){
    stop(paste("delta must be in [",minmax[1],",",minmax[2],"]", "\n", ""))
  }
  
  X<-tryCatch(cbind(rep(0,n), delta), warning=function(w) {
    stop("Input vectors have incompatible lengths")
    })

  if(!rot%in%c(0,90,180,270)){
    stop(paste("rotation must be in {0, 90, 180, 270}", "\n", ""))
  }
  
  N<-n
  out<-lapply(1:N, function(i) rcop_copula(delta=X[i,2, drop=T], distr=distr))
  out<-matrix(unlist(out), byrow=TRUE, nrow=N)

  return(out)
}

pcop_copula<-function(W, delta=0, distr="normal", rot=0, log.p = FALSE){
  #dcop wrapper function for distribution function for copula with scalar inputs

  if(distr=="independent"){
    cop_object<-copula::indepCopula(param=0, dim = 2)
  }

  if(distr=="normal"){
    #cop_dim<-nrow(copula::p2P(c(delta)))
    cop_object<-copula::normalCopula(param=delta, dim = 2)
  }

  if(distr=="clayton"){
    cop_object<-copula::claytonCopula(param=delta, dim = 2)
  }

  if(distr=="gumbel"){
    cop_object<-copula::gumbelCopula(param=delta, dim = 2)
  }

  if(distr=="frank"){
    cop_object<-copula::frankCopula(param=delta, dim = 2)
  }

  if(distr=="joe"){
    cop_object<-copula::joeCopula(param=delta, dim = 2)
  }

  if(distr=="amh"){
    cop_object<-copula::amhCopula(param=delta, dim = 2)
  }
  
  if(rot==90){
    cop_object<-copula::rotCopula(cop_object, flip = c(TRUE, FALSE))
  }
  
  if(rot==180){
    cop_object<-copula::rotCopula(cop_object, flip = c(TRUE, TRUE))
  }
  
  if(rot==270){
    cop_object<-copula::rotCopula(cop_object, flip = c(FALSE, TRUE))
  }
  
  
  #Wrapper for Copula package function
  out<-copula::pCopula(u=W, copula=cop_object)

  if(log.p){
    out<-log(out)
  }

  return(out)
}


rcop_copula<-function(delta=0, distr="normal", rot=0){
  #wrapper function for random number generation for copula with scalar inputs

  if(distr=="independent"){
    cop_object<-copula::indepCopula(param=0, dim = 2)
  }

  if(distr=="normal"){
    #cop_dim<-nrow(copula::p2P(c(delta)))
    cop_object<-copula::normalCopula(param=delta, dim = 2)
  }

  if(distr=="clayton"){
    cop_object<-copula::claytonCopula(param=delta, dim = 2)
  }

  if(distr=="gumbel"){
    cop_object<-copula::gumbelCopula(param=delta, dim = 2)
  }

  if(distr=="frank"){
    cop_object<-copula::frankCopula(param=delta, dim = 2)
  }

  if(distr=="joe"){
    cop_object<-copula::joeCopula(param=delta, dim = 2)
  }

  if(distr=="amh"){
    cop_object<-copula::amhCopula(param=delta, dim = 2)
  }
  
  if(rot==90){
    cop_object<-copula::rotCopula(cop_object, flip = c(TRUE, FALSE))
  }
  
  if(rot==180){
    cop_object<-copula::rotCopula(cop_object, flip = c(TRUE, TRUE))
  }
  
  if(rot==270){
    cop_object<-copula::rotCopula(cop_object, flip = c(FALSE, TRUE))
  }
  
  #Wrapper for Copula package function
  out<-copula::rCopula(n=1, copula=cop_object)

  return(out)
}
